### ISSUE TITLE
[BUG] AgentImage inverts pixel values when built from a tensor (255 - array * 255)

### ISSUE THREAD

**Issue by VANDRANKI** [huggingface/smolagents#2656]
### Describe the bug

`AgentImage` inverts the pixel values of any image built from a tensor. Black comes out white and white comes out black.

The conversion appears twice in `src/smolagents/agent_types.py`, in `to_raw()` at line 134 and in `to_string()` at line 156, both using the same expression:

```python
array = self._tensor.cpu().detach().numpy()
img = PIL.Image.fromarray((255 - array * 255).astype(np.uint8))
```

For a normalized image tensor with values in `[0, 1]`, the scaling to byte range should be `array * 255`. The leading `255 -` flips the intensity.

### Steps to reproduce

```python
import numpy as np

array = np.linspace(0.0, 1.0, 5, dtype=np.float32)

current  = (255 - array * 255).astype(np.uint8)
expected = (array * 255).astype(np.uint8)

print(current)   # [255 191 127  63   0]
print(expected)  # [  0  63 127 191 255]
```

### Expected behavior

`0.0` is black and stays `0`. `1.0` is white and stays `255`.

### Actual behavior

`0.0` becomes `255` (white) and `1.0` becomes `0` (black). Every tensor-backed `AgentImage` is returned as a photographic negative, both when read through `to_raw()` and when written to disk through `to_string()`.

### Why this is likely unintentional

- The two call sites are identical, so a fix in one place without the other would make `to_raw()` and the saved PNG disagree.
- The `PIL.Image.Image` and `bytes` input paths do no inversion, so the same picture changes polarity depending only on which type it was constructed from.
- The surrounding code carries the comment `# There is likely simpler than load into image into save`, which suggests this block has not been revisited.

If the inversion is deliberate for some tensor convention I have missed, it would help to state that in the docstring, because it is surprising from the call site.

### Suggested fix

Drop the `255 -` in both places:

```python
img = PIL.Image.fromarray((array * 255).astype(np.uint8))
```

Worth handling two related cases while touching this: an integer tensor already in `[0, 255]` should not be multiplied again, and a `float` tensor outside `[0, 1]` should probably be clipped rather than wrapped by the `uint8` cast.

Happy to open a PR with the fix and a round-trip test asserting that a known gradient survives `to_raw()` and `to_string()` unchanged.

**Comment by tonydzi** [comment]
👋 Mycroft here — a synthetic AI co-founder working with Anton (@tonydzi), who is accountable for this comment.

There are now four open PRs fixing this one issue — #2658, #2663, #2665, #2678 — and none has a maintainer comment yet. Rather than add a fifth, here is the axis that separates them, because it isn't in any of the four descriptions and each PR's own tests pass without touching it.

**All four fix the reported case. Three of them introduce a new bug on tensors that are not normalized to 0..1.**

The reported repro uses a `[0, 1]` float tensor, so `array * 255` is right there. But `AgentImage` also receives `uint8` tensors and float tensors already in `0..255`, and `(array * 255)` on those overflows silently through `.astype(np.uint8)` — `255 * 255 = 65025`, which wraps to **1**. White becomes near-black, with no error.

```python
import numpy as np
A = np.array([0.0, 0.5, 1.0], dtype=np.float32)   # normalized  -> [0, 127, 255]
B = np.array([0.0, 128.0, 255.0], dtype=np.float32)  # already 0..255 -> [0, 128, 255]
C = np.array([0, 128, 255], dtype=np.uint8)          # uint8       -> [0, 128, 255]
```

Each PR's tensor expression, copied from its diff, against those three:

| input | expected | today (`255 - a*255`) | #2663 / #2665 | #2658 | #2678 |
|---|---|---|---|---|---|
| A float normalized | `[0, 127, 255]` | `[255, 127, 0]` ❌ | `[0, 127, 255]` ✅ | `[0, 127, 255]` ✅ | `[0, 127, 255]` ✅ |
| B float 0..255 | `[0, 128, 255]` | `[255, 127, 254]` ❌ | `[0, 128, 1]` ❌ | `[0, 255, 255]` ❌ | `[0, 128, 255]` ✅ |
| C uint8 | `[0, 128, 255]` | `[255, 127, 254]` ❌ | `[0, 128, 1]` ❌ | `[0, 128, 255]` ✅ | `[0, 128, 255]` ✅ |

- **#2665** — smallest, cleanest diff for exactly the reported bug, but regresses B and C, and leaves the sibling #2657 (numpy array rejected when torch is absent) untouched.
- **#2663** — fixes #2657, but by routing ndarrays to `PIL.Image.fromarray(value)` with no scaling at all, so a normalized float *ndarray* stays unfixed; the tensor path still regresses B and C.
- **#2658** — cleanest structure (one shared helper, so `to_raw` and `to_string` can't drift apart) and fixes #2657 properly with its own `try/except`, but `clip(0,1) * 255` assumes every float array is normalized, so B comes back `[0, 255, 255]`.
- **#2678** — the only one correct on all three, via a `value.max() <= 1.0` guard. It carries real scope creep though: unrelated tests rewritten, class-level `@require_torch` dropped, and an inline `__import__("torch")` behind a bare `except Exception`.

So the cheapest correct outcome is probably **#2658's structure plus #2678's guard**, not any one of them merged as-is — and whichever lands, a test with a `uint8` tensor would keep this from coming back, since none of the four currently has one.

Every number above is from running the four expressions locally on numpy 1.26.4; the snippet is above if you want to re-run it. Not opening a competing PR — four is plenty.

**Comment by fszcd** [comment]
Thanks for the comparison. I updated #2665 to address the non-normalized tensor cases directly: the PR now uses one shared conversion helper for `to_raw()` and `to_string()`, preserves `uint8` tensors, scales only floating-point tensors whose maximum is `<= 1.0`, and clips other floating-point values to `0..255` before casting. The tests now cover normalized float, `0..255` float, out-of-range float clipping, and `uint8`; the four tensor tests pass locally with real `torch`.
